### ISSUE TITLE
flutter_blue not working on Android

### ISSUE THREAD

**Issue by juliansteenbakker** [rmawatson/flutter_isolate#71]
I have tried with different versions of flutter_isolate and flutter_blue however i keep getting the following error on Android. I have tried different Android devices ranging from Samsung Galaxy S9 to Galaxy Tab S2.

```dart
E/flutter (14042): [ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: MissingPluginException(No implementation found for method isAvailable on channel plugins.pauldemarco.com/flutter_blue/methods)
E/flutter (14042): #0      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:156:7)
E/flutter (14042): <asynchronous suspension>
E/flutter (14042): #1      FlutterBlue.isAvailable.<anonymous closure> (package:flutter_blue/src/flutter_blue.dart)
E/flutter (14042): <asynchronous suspension>
E/flutter (14042): 
I/flutter (14042): Error starting scan.
E/flutter (14042): [ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: MissingPluginException(No implementation found for method startScan on channel plugins.pauldemarco.com/flutter_blue/methods)
E/flutter (14042): #0      FlutterBlue.scan (package:flutter_blue/src/flutter_blue.dart:124:7)
E/flutter (14042): <asynchronous suspension>
E/flutter (14042): 
```

**Comment by nmfisher** [comment]
I'll look into this - have you tried the PR @ https://github.com/rmawatson/flutter_isolate/pull/32? Adding a custom plugin registrant may fix, I'm not sure.

**Comment by nmfisher** [comment]
@juliansteenbakker I have merged some fixes into master that allow for a custom plugin registrant, can you try this and see if it fixes?

**Comment by juliansteenbakker** [comment]
Sorry for late response, i have tried using the `customPluginRegistrant` however it keeps saying:
```
W/FlutterEngineCxnRegstry(18810): Attempted to register plugin (com.pauldemarco.flutter_blue.FlutterBluePlugin@f34dea1) but it was already registered with this FlutterEngine (io.flutter.embedding.engine.FlutterEngine@6de08c6).
I/FlutterIsolate(18810): Using custom Flutter plugin registrant com.xxx.xxx.CustomPluginRegistrant
```

I tried removing it from the `generatedPluginRegistrant` however it keeps coming back. I have set-up everything as provided in the example.

**Comment by nmfisher** [comment]
@juliansteenbakker after looking into this further, the problem is that on Android, the flutter_blue plugin requires an Activity to function (and flutter_isolate works by spawning a new FlutterEngine which isn't attached to one). In other words, any plugin called from an isolate won't be able to access the main Flutter Activity. On Android, this is needed to do certain things like request Bluetooth permissions, which is why flutter_blue will fail when invoked from a spawned isolate.

This is the same problem I encountered in the permissions_handler plugin. It's technically possible for the plugin author to fix (by launching a temporary Activity whenever permissions are being requested). We might be able to handle this on the flutter_isolate side by launching an Activity for the lifetime of each spawned FlutterEngine. I'd need to look into this further to see what the implications are.

In short, this might be fixable but I wouldn't hold my breath for any fixes in the short-term. The only short-term solution is to move your flutter_blue work to the main Flutter isolate (which is what I did with my permissions_handler work), or to fork flutter_blue and try and launch an Activity yourself.

(P.S. the "already registered messages" are a red herring, they can be easily fixed but that's not the source of the problem).

**Comment by campcoxconsultants** [comment]
Why can't I make sure I have the permissions in the main Flutter isolate and then run Flutter blue where it won't need an activity?

**Comment by nmfisher** [comment]
@campcoxconsultants I vaguely recall `flutter_blue` running a permissions check  against the activity every time a scan was performed (which would throw an exception, because the activity was null). It's been a while, though, so I don't remember exactly - did you manage to try this?

**Comment by calebadavis** [comment]
So if we were to tweak flutter_blue (or now flutter_blue_plus) to not perform the permissions check, and simply do that in the main Isolate instead (using a custom MethodChannel), perhaps we could get around this? I'll poke around the bluetooth library and see if the permissions calls are isolated enough to just comment out...

**Comment by yeonsann** [comment]
@calebadavis Any update on this?

**Comment by calebadavis** [comment]
@yeonsann Yes, I took a look, I believe it's possible. In the bowels of FlutterBluePlus, In the Android plugin code, file 'FlutterBluePlusPlugin.java', line 670 (method 'ensurePermissionBeforeAction()'), we can comment everything out and simply always call 'operation.op(true, permission);'. This will bypass permission checks.

Additionally, somewhere earlier in the flutter app, you need a custom method channel calling into Android code to actually do the check for / request permission at least once (in the init code for a stateful widget). 

I have this custom channel working in my own app for different reasons (not in an isolate). 

Google has pretty thorough documentation on how to write these things: https://docs.flutter.dev/development/platform-integration/platform-channels

**Comment by oalinoor11** [comment]
A problem occurred configuring project ':flutter_blue'.
> Could not get unknown property 'source' for generate-proto-generateDebugProto of type org.gradle.api.internal.file.DefaultSourceDirectorySet.

**Comment by GPlay97** [comment]
any updates? Impossible to use with good performance, when bluetooth reading blocks the whole main thread
